File: DWSIM.UI.Desktop.Forms/Forms/AboutBox.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Eto.Forms;
using Eto.Drawing;
using System.Reflection;
using System.IO;

using DWSIM.UI.Shared;
using System.Collections.ObjectModel;

namespace DWSIM.UI.Forms.Forms
{
    class AboutBox: Form
    {

        private double sf = GlobalSettings.Settings.UIScalingFactor;

        private ObservableCollection<ComponentInfo> components = new ObservableCollection<ComponentInfo>();
        private ObservableCollection<ComponentInfo> componentsn = new ObservableCollection<ComponentInfo>();

        public AboutBox(): base()
        {
            Init();
        }

        public void Init()
        {

            int w = (int)(sf * 640);
            int h = (int)(sf * 480);

            var center = Screen.PrimaryScreen.WorkingArea.Center;
            center.X -= w / 2;
            center.Y -= h / 2;

            Location = new Point(center);

            ClientSize = new Size(w, h);
            
            Maximizable = false;

            Minimizable = false;

            Resizable = false;

            ShowInTaskbar = false;

            AddComponentInfo();
            AddComponentInfoN();

            string imgprefix = "DWSIM.UI.Forms.Resources.Icons.";

            Title = "AboutDWSIM".Localize();

            Icon = Eto.Drawing.Icon.FromResource(imgprefix + "DWSIM_ico.ico");

            var layout = new PixelLayout();

            string vtext = "Version".Localize() + " " + Assembly.GetExecutingAssembly().GetName().Version.Major.ToString() + 
                "." + Assembly.GetExecutingAssembly().GetName().Version.Minor.ToString() +
                "." + Assembly.GetExecutingAssembly().GetName().Version.Build.ToString();

#if DEBUG
            vtext += "-" + File.GetLastWriteTimeUtc(Assembly.GetExecutingAssembly().Location).ToString();
#endif

            string crtext = Shared.AssemblyCopyright;

            layout.Add(new ImageView { Size = new Size((int)(sf * 100), (int)(sf * 100)), Image = new Bitmap(Eto.Drawing.Bitmap.FromResource(imgprefix + "DWSIM_ico.png")) }, 0, 0);
            layout.Add(new Label { Text = "DWSIM Simulator (Cross-Platform User Interface)", TextAlignment = TextAlignment.Left, Font = SystemFonts.Bold(null, FontDecoration.None) }, (int)(sf * 110), 0);
            layout.Add(new Label { Text = vtext, TextAlignment = TextAlignment.Left}, (int)(sf * 110), (int)(sf * 20));
            layout.Add(new Label { Text = crtext, TextAlignment = TextAlignment.Left}, (int)(sf * 110), (int)(sf * 40));

            string osinfo = "", clrinfo = "", meminfo = "";

            if (Application.Instance.Platform.IsMac)
            {
                var osversion = Environment.OSVersion.Version.ToString();
                switch (osversion)
                {
                    case "11.0.0.0": osinfo = "Mac OS X Lion (v10.7.0)"; break;
                    case "11.4.2.0": osinfo = "Mac OS X Lion (v10.7.5)"; break;
                    case "12.0.0.0": osinfo = "OS X Mountain Lion (v10.8.0)"; break;
                    case "13.0.0.0": osinfo = "OS X Mavericks (v10.9.0)"; break;
                    case "13.4.0.0": osinfo = "OS X Mavericks (v10.9.5)"; break;
                    case "14.0.0.0": osinfo = "OS X Yosemite (v10.10.0)"; break;
                    case "14.5.0.0": osinfo = "OS X Yosemite (v10.10.5)"; break;
                    case "15.0.0.0": osinfo = "OS X El Captain (v10.11.0)"; break;
                    case "15.6.0.0": osinfo = "OS X El Captain (v10.11.6)"; break;
                    case "16.0.0.0": osinfo = "macOS Sierra (v10.12.0)"; break;
                    case "16.1.0.0": osinfo = "macOS Sierra (v10.12.1)"; break;
                    case "16.3.0.0": osinfo = "macOS Sierra (v10.12.2)"; break;
                    case "16.4.8.0": osinfo = "macOS Sierra (v10.12.3)"; break;
                    case "16.5.3.0": osinfo = "macOS Sierra (v10.12.4)"; break;
                    case "16.5.6.0": osinfo = "macOS Sierra (v10.12.5)"; break;
                    case "16.7.1.2": osinfo = "macOS Sierra (v10.12.6)"; break;
                    case "17.0.0.0": osinfo = "macOS High Sierra (v10.13.0)"; break;
                    case "17.4.0.0": osinfo = "macOS High Sierra (v10.13.3)"; break;
                    case "18.0.3.0": osinfo = "macOS Mojave (v10.14.0)"; break;
                    case "18.3.0.0": osinfo = "macOS Mojave (v10.14.1-3)"; break;
                    case "18.5.0.0": osinfo = "macOS Mojave (v10.14.4)"; break;
                    case "18.6.0.0": osinfo = "macOS Mojave (v10.14.5)"; break;
                    case "18.7.1.1": osinfo = "macOS Mojave (v10.14.6)"; break;
                    case "19.0.0.0": osinfo = "macOS Catalina (v10.15)"; break;
                    case "19.2.0.0": osinfo = "macOS Catalina (v10.15.2)"; break;
                    case "19.3.0.0": osinfo = "macOS Catalina (v10.15.3)"; break;
                    case "19.4.0.0": osinfo = "macOS Catalina (v10.15.4)"; break;
                    case "19.5.0.0": osinfo = "macOS Catalina (v10.15.5)"; break;
                    case "19.6.0.0": osinfo = "macOS Catalina (v10.15.7)"; break;
                    case "20.0.0.0": osinfo = "macOS Big Sur (v11.0.0)"; break;
                    case "20.1.0.0": osinfo = "macOS Big Sur (v11.0.1)"; break;
                    case "20.3.0.0": osinfo = "macOS Big Sur (v11.1.0)"; break;
                    default: osinfo = "macOS (v" + osversion + ")"; break;
                }
            }
            else {
                osinfo = Environment.OSVersion.ToString();
            }

            clrinfo = SharedClasses.Utility.GetRuntimeVersion();
            
            meminfo = (GC.GetTotalMemory(false) / 1024 / 1024).ToString("#") + " MB managed, " + (Environment.WorkingSet / 1024 / 1024).ToString("#") + " MB total";

            var container1 = new DynamicLayout() { Padding = new Padding(10) };

            container1.CreateAndAddDescriptionRow("DWSIM is released under the terms of the GNU General Public License (GPL) version 3.");
            container1.CreateAndAddTwoLabelsRow2("OS Info:", osinfo);
            container1.CreateAndAddTwoLabelsRow2("CLR Info:", clrinfo);
            container1.CreateAndAddTwoLabelsRow2("Memory Usage:", meminfo);
            
            var listcontainer = new GridView { DataStore = components, RowHeight = (int)(sf * 20) };
            var listcontainern = new GridView { DataStore = componentsn, RowHeight = (int)(sf * 20) };

            var col1 = new GridColumn
            {
                DataCell = new TextBoxCell { Binding = Binding.Property<ComponentInfo, string>(r => r.Name) },
                HeaderText = "Name"
            };
            col1.AutoSize = true;
            listcontainer.Columns.Add(col1);
            var col1a = new GridColumn
            {
                DataCell = new TextBoxCell { Binding = Binding.Property<ComponentInfo, string>(r => r.Version) },
                HeaderText = "Version"
            };
            col1a.AutoSize = true;
            listcontainer.Columns.Add(col1a);
            var col1b = new GridColumn
            {
                DataCell = new TextBoxCell { Binding = Binding.Property<ComponentInfo, string>(r => r.Year) },
                HeaderText = "Year"
            };
            col1b.AutoSize = true;
            listcontainer.Columns.Add(col1b);
            var col1c = new GridColumn
            {
                DataCell = new TextBoxCell { Binding = Binding.Property<ComponentInfo, string>(r => r.Copyright) },
                HeaderText = "Copyright"
            };
            col1c.AutoSize = true;
            listcontainer.Columns.Add(col1c);
            var col1d = new GridColumn
            {
                DataCell = new TextBoxCell { Binding = Binding.Property<ComponentInfo, string>(r => r.Website) },
                HeaderText = "Website"
            };
            col1d.AutoSize = true;
            listcontainer.Columns.Add(col1d);
            var col1e = new GridColumn
            {
                DataCell = new TextBoxCell { Binding = Binding.Property<ComponentInfo, string>(r => r.License) },
                HeaderText = "License"
            };
            col1e.AutoSize = true;
            listcontainer.Columns.Add(col1e);
            var col1f = new GridColumn
            {
                DataCell = new TextBoxCell { Binding = Binding.Property<ComponentInfo, string>(r => r.LicenseText) },
                HeaderText = "License Text"
            };
            col1f.AutoSize = true;
            listcontainer.Columns.Add(col1f);


            var col1n = new GridColumn
            {
                DataCell = new TextBoxCell { Binding = Binding.Property<ComponentInfo, string>(r => r.Name) },
                HeaderText = "Name"
            };
            col1n.AutoSize = true;
            listcontainern.Columns.Add(col1n);
            var col1an = new GridColumn
            {
                DataCell = new TextBoxCell { Binding = Binding.Property<ComponentInfo, string>(r => r.Version) },
                HeaderText = "Version"
            };
            col1an.AutoSize = true;
            listcontainern.Columns.Add(col1an);
            var col1fn = new GridColumn
            {
                DataCell = new TextBoxCell { Binding = Binding.Property<ComponentInfo, string>(r => r.LicenseText) },
                HeaderText = "License Text"
            };
            col1fn.AutoSize = true;
            listcontainern.Columns.Add(col1fn);
            
            string gpltext;
            using (Stream stream = Assembly.GetExecutingAssembly().GetManifestResourceStream("DWSIM.UI.Forms.Resources.TextFiles.gpl-3.0.txt"))
            using (StreamReader reader = new StreamReader(stream))
            {
                gpltext = reader.ReadToEnd();
            }

            var txt1 = new TextArea { Text = gpltext, ReadOnly = true, Font = Fonts.Monospace(10.0f) };

            var t1 = new TableLayout(new TableRow(txt1));

            var tab1 = new TabPage { Content = container1, Text = "General Information" };
            var tab2 = new TabPage { Content = new Scrollable { Content = t1 }, Text = "GPLv3 License" };
            var tab3 = new TabPage { Content = new Scrollable { Content = listcontainer }, Text = "External Components" };
            var tab4 = new TabPage { Content = new Scrollable { Content = listcontainern }, Text = "NuGet Packages" };

            var tabc = new TabControl();
            tabc.Pages.Add(tab1);
            tabc.Pages.Add(tab2);
            tabc.Pages.Add(tab3);
            tabc.Pages.Add(tab4);

            var tablecontainer = new TableLayout { Padding = new Padding(10), Spacing = new Size(5, 5) };

            tablecontainer.Rows.Add(new TableRow(layout));
            tablecontainer.Rows.Add(new TableRow(tabc));

            Content = tablecontainer;

        }

        private void AddComponentInfo()
        {
            components.Add(new ComponentInfo("CoolProp", "6.0.0", "2016", "Ian H. Bell", "http://wwww.coolprop.org", "MIT-style License", "https://github.com/ibell/coolprop/blob/master/LICENSE"));
            components.Add(new ComponentInfo("GERG 2008", "2.0", "2017", "E. W. Lemmon, V. Heinemann, J. Lu, I. Bell", "https://github.com/usnistgov/AGA8", "NIST/17 U.S.C. 105", "https://github.com/usnistgov/AGA8/blob/master/LICENSE"));
            components.Add(new ComponentInfo("ChemSep Database", "7.30", "2018", "Harry Kooijman, Ross Taylor", "http://www.chemsep.org", "Perl Artistic License v2", "http://www.perlfoundation.org/artistic_license_2_0"));
            components.Add(new ComponentInfo("Flee", "0.9.14", "2009", "Eugene Ciloci", "https://flee.codeplex.com", "LGPLv2", "http://www.gnu.org/licenses/lgpl.html"));
            components.Add(new ComponentInfo("DotNumerics", "1.0", "2009", "Jose Antonio De Santiago Castillo", "http://www.dotnumerics.com", "GPLv3", "http://www.gnu.org/licenses/gpl.html"));
            components.Add(new ComponentInfo("FileHelpers", "3.2.7", "2018", "Marcos Meli", "https://sourceforge.net/projects/filehelpers", "LGPLv2", "http://www.gnu.org/licenses/lgpl.html"));
            components.Add(new ComponentInfo("SharpZipLib", "0.85.4.369", "2010", "IC#Code", "http://www.icsharpcode.net/OpenSource/SharpZipLib", "GPLv2", "http://www.gnu.org/licenses/gpl.html"));
            components.Add(new ComponentInfo("Nini", "1.1", "2010", "Brent R. Matzelle", "https://sourceforge.net/projects/nini", "MIT License", "http://www.opensource.org/licenses/mit-license.html"));
            components.Add(new ComponentInfo("ScintillaNET", "3.5.1.0", "2015", "Jacob Slusser", "https://github.com/jacobslusser/scintillaNET", "MIT License", "http://www.opensource.org/licenses/mit-license.html"));
            components.Add(new ComponentInfo("Yeppp!", "1.0.0.1", "2014", "Marat Dukhan", "http://www.yeppp.info", "Yeppp! License", "http://www.yeppp.info/resources/yeppp-license.txt"));
            components.Add(new ComponentInfo("SwarmOps", "3.1", "2011", "Magnus Erik Hvass Pedersen", "http://www.hvass-labs.org/projects/swarmops/cs/", "MIT-style License", "http://www.hvass-labs.org/projects/swarmops/cs/files/license.txt"));
            components.Add(new ComponentInfo("RandomOps", "2.1", "2010", "Magnus Erik Hvass Pedersen", "http://www.hvass-labs.org/projects/randomops/cs/", "MIT-style License", "http://www.hvass-labs.org/projects/randomops/cs/files/license.txt" ));
            components.Add(new ComponentInfo("OxyPlot", "2.0", "2017", "OxyPlot team", "http://www.oxyplot.org", "MIT License", "http://www.opensource.org/licenses/mit-license.html"));
            components.Add(new ComponentInfo("SkiaSharp", "1.68.0", "2020", "Xamarin team", "https://github.com/mono/SkiaSharp/", "MIT License", "https://github.com/mono/SkiaSharp/blob/master/LICENSE.md"));
            components.Add(new ComponentInfo("Eto.Forms", "custom version", "2019", "", "https://github.com/DanWBR/Eto", "BSD-style License", "https://github.com/DanWBR/Eto/blob/develop/LICENSE.txt"));
            components.Add(new ComponentInfo("Eto.Platform.Gtk2", "custom version", "2019", "", "https://github.com/DanWBR/Eto", "BSD-style License", "https://github.com/DanWBR/Eto/blob/develop/LICENSE.txt"));
            components.Add(new ComponentInfo("Eto.Platform.Windows", "custom version", "2019", "", "https://github.com/DanWBR/Eto", "BSD-style License", "https://github.com/DanWBR/Eto/blob/develop/LICENSE.txt"));
            components.Add(new ComponentInfo("Eto.Platform.Wpf", "custom version", "2019", "", "https://github.com/DanWBR/Eto", "BSD-style License", "https://github.com/DanWBR/Eto/blob/develop/LICENSE.txt"));
            components.Add(new ComponentInfo("Eto.Platform.XamMac2", "custom version", "2019", "", "https://github.com/DanWBR/Eto", "BSD-style License", "https://github.com/DanWBR/Eto/blob/develop/LICENSE.txt"));
            components.Add(new ComponentInfo("ReoGrid", "custom version", "2020", "", "https://github.com/DanWBR/ReoGrid", "MIT-style License", "https://github.com/DanWBR/ReoGrid/blob/master/LICENSE"));
        }

        private void AddComponentInfoN()
        {
           
            componentsn.Add(new ComponentInfo("cef.redist.x64", "75.1.14", "https://raw.github.com/cefsharp/cef-binary/master/LICENSE.txt"));
            componentsn.Add(new ComponentInfo("cef.redist.x86", "75.1.14", "https://raw.github.com/cefsharp/cef-binary/master/LICENSE.txt"));
            componentsn.Add(new ComponentInfo("CefSharp.Common", "75.1.14", "https://raw.github.com/cefsharp/CefSharp/master/LICENSE"));
            componentsn.Add(new ComponentInfo("CefSharp.WinForms", "75.1.14", "https://raw.github.com/cefsharp/CefSharp/master/LICENSE"));
            componentsn.Add(new ComponentInfo("DynamicLanguageRuntime", "1.2.3", "https://github.com/IronLanguages/dlr/blob/master/LICENSE"));
            componentsn.Add(new ComponentInfo("IronPython", "2.7.10", "https://github.com/IronLanguages/ironpython2/blob/master/LICENSE"));
            componentsn.Add(new ComponentInfo("IronPython.StdLib", "2.7.10", "http://docs.python.org/license.html"));
            componentsn.Add(new ComponentInfo("jacobslusser.ScintillaNET", "3.6.3", ""));
            componentsn.Add(new ComponentInfo("MathNet.Numerics", "4.7.0", "https://numerics.mathdotnet.com/License.html"));
            componentsn.Add(new ComponentInfo("Newtonsoft.Json", "12.0.3", "https://raw.github.com/JamesNK/Newtonsoft.Json/master/LICENSE.md"));
            componentsn.Add(new ComponentInfo("OpenTK", "3.0.", "http://github.com/opentk/opentk/blob/master/License.txt"));
            componentsn.Add(new ComponentInfo("OpenTK.GLControl", "3.0.1", "http://github.com/opentk/opentk/blob/master/License.txt"));
            componentsn.Add(new ComponentInfo("OxyPlot.Core", "2.0.0-unstable0956", "https://raw.githubusercontent.com/oxyplot/oxyplot/master/LICENSE"));
            componentsn.Add(new ComponentInfo("OxyPlot.Wpf", "2.0.0-unstable0956", "https://raw.githubusercontent.com/oxyplot/oxyplot/master/LICENSE"));
            componentsn.Add(new ComponentInfo("SharpDX", "4.0.1", "http://sharpdx.org/License.txt"));
            componentsn.Add(new ComponentInfo("SharpDX.Direct2D1", "4.0.1", "http://sharpdx.org/License.txt"));
            componentsn.Add(new ComponentInfo("SharpDX.DXGI", "4.0.1", "http://sharpdx.org/License.txt"));
            componentsn.Add(new ComponentInfo("SharpDX.Mathematics", "4.0.1", "http://sharpdx.org/License.txt"));
            componentsn.Add(new ComponentInfo("SharpZipLib", "1.1.0", "https://github.com/icsharpcode/SharpZipLib/blob/master/LICENSE.txt"));
            componentsn.Add(new ComponentInfo("SkiaSharp", "1.68.2.1", "https://github.com/mono/SkiaSharp/blob/master/LICENSE.md"));
            componentsn.Add(new ComponentInfo("SkiaSharp.Extended", "1.68.2.1", "https://github.com/mono/SkiaSharp.Extended/blob/master/LICENSE"));
            componentsn.Add(new ComponentInfo("System.ComponentModel", "4.3.0", "http://go.microsoft.com/fwlink/?LinkId=329770"));
            componentsn.Add(new ComponentInfo("System.Runtime.Serialization.Primitives", "4.3.0", "http://go.microsoft.com/fwlink/?LinkId=329770"));
            componentsn.Add(new ComponentInfo("Eto.OxyPlot", "1.2.0-beta", ""));
            componentsn.Add(new ComponentInfo("Eto.OxyPlot.Wpf", "1.2.0-beta", ""));
        }

        public class ComponentInfo
        {
            public ComponentInfo(string name, string version, string year, string copyright, string website, string license, string licensetext)
            {
                Name = name;
                Version = version;
                Year = year;
                Copyright = copyright;
                Website = website;
                License = license;
                LicenseText = licensetext;
            }

            public ComponentInfo(string name, string version, string licensetext)
            {
                Name = name;
                Version = version;
                LicenseText = licensetext;
            }

            public string Name { get; set; }

            public string Version { get; set; }

            public string Year { get; set; }

            public string Copyright { get; set; }
            public string Website { get; set; }

            public string License { get; set; }

            public string LicenseText { get; set; }

        }

    }
}

File: DWSIM/Forms/AboutBox.vb
Imports System.Reflection
Imports System.Text.RegularExpressions
Imports System.IO
Imports Yeppp
Imports System.Linq
Imports Cudafy
Imports Cudafy.Host

Public Class AboutBox

    Private _IsPainted As Boolean = False
    Private _EntryAssemblyName As String
    Private _CallingAssemblyName As String
    Private _ExecutingAssemblyName As String
    Private _EntryAssembly As System.Reflection.Assembly
    Private _EntryAssemblyAttribCollection As Specialized.NameValueCollection

    Private Sub AboutBox_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

        ExtensionMethods.ChangeDefaultFont(Me)

        TextBox1.Font = New Font("Consolas", 9, GraphicsUnit.Point)
        tbAcknowledgements.Font = New Font("Consolas", 9, GraphicsUnit.Point)

        Dim updfile = My.Application.Info.DirectoryPath & Path.DirectorySeparatorChar & "version.info"

        Version.Text = "Version " & My.Application.Info.Version.Major & "." & My.Application.Info.Version.Minor & "." & My.Application.Info.Version.Build

#If DEBUG Then
        Version.Text += "-" + IO.File.GetLastWriteTimeUtc(Assembly.GetExecutingAssembly().Location).ToString()
#End If

        tbAcknowledgements.Text = "A HUGE thank you to the following Patrons/Sponsors who made this release possible:" + vbCrLf + vbCrLf +
            Patrons.GetList() + vbCrLf + vbCrLf + tbAcknowledgements.Text

        Copyright.Text = My.Application.Info.Copyright

        LblOSInfo.Text = My.Computer.Info.OSFullName & ", Version " & My.Computer.Info.OSVersion & ", " & My.Computer.Info.OSPlatform & " Platform"
        LblCLRInfo.Text = SharedClasses.Utility.GetRuntimeVersion()
        Lblmem.Text = (GC.GetTotalMemory(False) / 1024 / 1024).ToString("#") & " MB managed, " & (My.Application.Info.WorkingSet / 1024 / 1024).ToString("#") & " MB total"

        Lblcpuinfo.Text = "Retrieving CPU info..."

        If Not DWSIM.App.IsRunningOnMono Then

            Threading.Tasks.Task.Factory.StartNew(Function()
                                                      Dim scrh As New System.Management.ManagementObjectSearcher("select * from Win32_Processor")
                                                      Dim text1 As String = System.Environment.GetEnvironmentVariable("PROCESSOR_IDENTIFIER")
                                                      For Each qinfo In scrh.Get()
                                                          text1 += " / " & qinfo.Properties("Name").Value.ToString
                                                      Next
                                                      text1 += " (" & Yeppp.Library.GetProcessABI().Description & ")"
                                                      Return text1
                                                  End Function).ContinueWith(Sub(t)
                                                                                 Lblcpuinfo.Text = t.Result
                                                                             End Sub, Threading.Tasks.TaskScheduler.FromCurrentSynchronizationContext)

        Else

            Threading.Tasks.Task.Factory.StartNew(Function()
                                                      Dim sinfo As New ProcessStartInfo With {.FileName = "lshw", .Arguments = "-c CPU", .RedirectStandardOutput = True, .UseShellExecute = False}
                                                      Dim p As New Process With {.StartInfo = sinfo}
                                                      p.Start()
                                                      Dim output As String = p.StandardOutput.ReadToEnd
                                                      p.WaitForExit()
                                                      Dim lbltext As String = ""
                                                      For Each l In output.Split(New Char() {vbCrLf, vbLf, vbCr})
                                                          If l.Contains("product") Then
                                                              lbltext = l.Split(":")(1).TrimStart(" ")
                                                          End If
                                                          If l.Contains("vendor") Then
                                                              lbltext += " / " & l.Split(": ")(1).TrimStart(" ")
                                                              Exit For
                                                          End If
                                                      Next
                                                      Return lbltext
                                                  End Function).ContinueWith(Sub(t)
                                                                                 Lblcpuinfo.Text = t.Result
                                                                                 Lblcpuinfo.Text += " (" & Yeppp.Library.GetProcessABI().Description & ")"
                                                                             End Sub, Threading.Tasks.TaskScheduler.FromCurrentSynchronizationContext)

        End If

        Lblcpusimd.Text = "Querying CPU SIMD capabilities..."

        Threading.Tasks.Task.Factory.StartNew(Function()
                                                  Dim text1 As String = ""
                                                  For Each item In Library.GetCpuArchitecture.CpuSimdFeatures
                                                      text1 += item.ToString & " "
                                                  Next
                                                  Return text1
                                              End Function).ContinueWith(Sub(t)
                                                                             Lblcpusimd.Text = t.Result
                                                                         End Sub, Threading.Tasks.TaskScheduler.FromCurrentSynchronizationContext)

        With Me.DataGridView1.Rows
            .Clear()
            .Add(New Object() {"GERG 2008", "2.0", "2017", "E. W. Lemmon, V. Heinemann, J. Lu, I. Bell", "https://github.com/usnistgov/AGA8", "NIST/17 U.S.C. 105", "https://github.com/usnistgov/AGA8/blob/master/LICENSE"})
            .Add(New Object() {"pyeq2", "10.1", "2013", "James R. Phillips", "https://code.google.com/p/pyeq2/", "BSD 3", "http://opensource.org/licenses/BSD-3-Clause"})
            .Add(New Object() {"IPOPT", "3.9.2", "2011", "COIN-OR", "https://projects.coin-or.org/Ipopt", "Eclipse Public License", "http://www.eclipse.org/legal/epl-v10.html"})
            .Add(New Object() {"lp_solve", "5.5", "2009", "Michel Berkelaar, Kjell Eikland, Peter Notebaert", "http://lpsolve.sourceforge.net", "LGPLv2", "http://www.gnu.org/licenses/lgpl.html"})
            .Add(New Object() {"CoolProp", "6.0.0", "2016-2018", "Ian H. Bell", "http://wwww.coolprop.org", "MIT-style License", "https://github.com/ibell/coolprop/blob/master/LICENSE"})
            .Add(New Object() {"ChemSep Database", "8.01", "2018", "Harry Kooijman, Ross Taylor", "http://www.chemsep.org", "Perl Artistic License v2", "http://www.perlfoundation.org/artistic_license_2_0"})
            .Add(New Object() {"Flee", "0.9.14", "2009", "Eugene Ciloci", "https://flee.codeplex.com", "LGPLv2", "http://www.gnu.org/licenses/lgpl.html"})
            .Add(New Object() {"CUDAfy", "1.25.4963.10126", "2013", "Hybrid DSP", "https://cudafy.codeplex.com", "LGPLv2", "http://www.gnu.org/licenses/lgpl.html"})
            .Add(New Object() {"DotNumerics", "1.0", "2009", "Jose Antonio De Santiago Castillo", "http://www.dotnumerics.com", "GPLv3", "http://www.gnu.org/licenses/gpl.html"})
            .Add(New Object() {"CSIPOPT", "1.0", "2012", "Anders Gustafsson, Cureos AB", "https://github.com/cureos/csipopt", "Eclipse Public License", "http://www.eclipse.org/legal/epl-v10.html"})
            .Add(New Object() {"NetOffice", "1.6", "2011", "Sebastian Lange", "https://netoffice.codeplex.com/", "MIT License", "https://netoffice.codeplex.com/license"})
            .Add(New Object() {"GemBox.Spreadsheet", "39.3.30.1037", "2015", "GemBox Software", "http://www.gemboxsoftware.com/spreadsheet/overview", "EULA", "http://www.gemboxsoftware.com/Spreadsheet/Eula.rtf"})
            .Add(New Object() {"Indigo", "1.1", "2013", "GGA Software Services LLC", "http://www.ggasoftware.com/opensource/indigo", "GPLv3", "http://www.gnu.org/licenses/gpl.html"})
            .Add(New Object() {"Nini", "1.1", "2010", "Brent R. Matzelle", "https://sourceforge.net/projects/nini", "MIT License", "http://www.opensource.org/licenses/mit-license.html"})
            .Add(New Object() {"Jolt.NET", "0.4", "2009", "Steve Guidi", "https://github.com/jacobslusser/scintillaNET", "New BSD License (BSD)", "http://jolt.codeplex.com/license"})
            .Add(New Object() {"Yeppp!", "1.0.0.1", "2014", "Marat Dukhan", "http://www.yeppp.info", "Yeppp! License", "http://www.yeppp.info/resources/yeppp-license.txt"})
            .Add(New Object() {"ExcelDNA", "0.34", "2017", "Govert van Drimmelen", "http://excel-dna.net/", "MIT License", "http://www.opensource.org/licenses/mit-license.html"})
            .Add(New Object() {"AODL", "1.4.0.3", "2011", "Chris Constantin", "https://bitbucket.org/chrisc/aodl", "Apache License v2", "https://wiki.openoffice.org/wiki/OpenOffice.org_Wiki:Copyrights"})
            .Add(New Object() {"Eto.Forms", "custom version", "2019", "", "https://github.com/DanWBR/Eto", "BSD-style License", "https://github.com/DanWBR/Eto/blob/develop/LICENSE.txt"})
            .Add(New Object() {"Eto.Platform.Gtk2", "custom version", "2019", "", "https://github.com/DanWBR/Eto", "BSD-style License", "https://github.com/DanWBR/Eto/blob/develop/LICENSE.txt"})
            .Add(New Object() {"Eto.Platform.Windows", "custom version", "2019", "", "https://github.com/DanWBR/Eto", "BSD-style License", "https://github.com/DanWBR/Eto/blob/develop/LICENSE.txt"})
            .Add(New Object() {"Eto.Platform.Wpf", "custom version", "2019", "", "https://github.com/DanWBR/Eto", "BSD-style License", "https://github.com/DanWBR/Eto/blob/develop/LICENSE.txt"})
            .Add(New Object() {"Eto.Platform.XamMac2", "custom version", "2019", "", "https://github.com/DanWBR/Eto", "BSD-style License", "https://github.com/DanWBR/Eto/blob/develop/LICENSE.txt"})
            .Add(New Object() {"ReoGrid", "custom version", "2019", "", "https://github.com/DanWBR/ReoGrid", "MIT-style License", "https://github.com/DanWBR/ReoGrid/blob/master/LICENSE"})
        End With
        Me.DataGridView1.Sort(Me.DataGridView1.Columns(0), System.ComponentModel.ListSortDirection.Ascending)

        With Me.DataGridView2.Rows
            .Clear()
            .Add(New Object() {"DynamicLanguageRuntime", "1.2.3", "https://github.com/IronLanguages/dlr/blob/master/LICENSE"})
            .Add(New Object() {"IronPython", "2.7.10", "https://github.com/IronLanguages/ironpython2/blob/master/LICENSE"})
            .Add(New Object() {"IronPython.StdLib", "2.7.10", "http://docs.python.org/license.html"})
            .Add(New Object() {"jacobslusser.ScintillaNET", "3.6.3", ""})
            .Add(New Object() {"MathNet.Numerics", "4.7.0", "https://numerics.mathdotnet.com/License.html"})
            .Add(New Object() {"Newtonsoft.Json", "12.0.3", "https://raw.github.com/JamesNK/Newtonsoft.Json/master/LICENSE.md"})
            .Add(New Object() {"OpenTK", "3.0.", "https://github.com/opentk/opentk/blob/master/License.txt"})
            .Add(New Object() {"OpenTK.GLControl", "3.0.1", "http://github.com/opentk/opentk/blob/master/License.txt"})
            .Add(New Object() {"OxyPlot.Core", "2.0.0-unstable0956", "https://raw.githubusercontent.com/oxyplot/oxyplot/master/LICENSE"})
            .Add(New Object() {"OxyPlot.Wpf", "2.0.0-unstable0956", "https://raw.githubusercontent.com/oxyplot/oxyplot/master/LICENSE"})
            .Add(New Object() {"SharpDX", "4.0.1", "http://sharpdx.org/License.txt"})
            .Add(New Object() {"SharpDX.Direct2D1", "4.0.1", "http://sharpdx.org/License.txt"})
            .Add(New Object() {"SharpDX.DXGI", "4.0.1", "http://sharpdx.org/License.txt"})
            .Add(New Object() {"SharpDX.Mathematics", "4.0.1", "http://sharpdx.org/License.txt"})
            .Add(New Object() {"SharpZipLib", "1.1.0", "https://github.com/icsharpcode/SharpZipLib/blob/master/LICENSE.txt"})
            .Add(New Object() {"SkiaSharp", "1.68.2.1", "https://github.com/mono/SkiaSharp/blob/master/LICENSE.md"})
            .Add(New Object() {"SkiaSharp.Extended", "1.68.2.1", "https://github.com/mono/SkiaSharp.Extended/blob/master/LICENSE"})
            .Add(New Object() {"System.ComponentModel", "4.3.0", "http://go.microsoft.com/fwlink/?LinkId=329770"})
            .Add(New Object() {"System.Runtime.Serialization.Primitives", "4.3.0", "http://go.microsoft.com/fwlink/?LinkId=329770"})
        End With

    End Sub

    Private Sub Button1_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Button1.Click
        Me.Close()
    End Sub


    ''' <summary>
    ''' populate a listview with the specified key and value strings
    ''' </summary>
    Private Sub Populate(ByVal lvw As ListView, ByVal Key As String, ByVal Value As String)
        If Value = "" Then Return
        Dim lvi As New ListViewItem
        lvi.Text = Key
        lvi.SubItems.Add(Value)
        lvw.Items.Add(lvi)
    End Sub

    Private Sub AboutBox_Shown(sender As Object, e As EventArgs) Handles Me.Shown
        FormMain.TranslateFormFunction?.Invoke(Me)
    End Sub
End Class